torch/utils/data/_utils/worker.py

```diff
@@ -220,9 +221,22 @@ def _worker_loop(dataset_kind, dataset, index_queue, data_queue, done_event,
         signal_handling._set_worker_signal_handlers()
 
         torch.set_num_threads(1)
+
+        # Note [RNG re-seeding in Dataloader workers]
+        # We need to explicitly re-seed all RNGs here (builtin random, numpy, torch) to ensure that the RNGs are
+        # different across workers.
+        # We have to seed the default RNG and the user-made ones differently otherwise the RNG of all Generators
+        # would be the same as the default one.
+        # For each Generator we have to draw a seed that depends on its current state (done in _generate_seed()),
+        # otherwise all Generator instances within a given worker would have the same RNG.
         seed = base_seed + worker_id
         random.seed(seed)
-        torch.manual_seed(seed)
+        generators = [o for o in gc.get_objects() if isinstance(o, torch.Generator)]
+        for generator in generators:
+            if generator is torch.random.default_generator:
+                torch.manual_seed(seed)
+            elif generator.device.type != "cuda":
+                generator.manual_seed(_generate_seed(generator) + seed)
```

Comment:
The fact that we have to seed the default RNG differently from the other ones is a bit ugly / brittle.
There is also this seemingly `+ seed` part which doesn't look like much but it's absolutely necessary to get different RNG across epochs.

I think it'd be cleaner if we could draw all the base seeds (one per Generator instance) in `dataloaer.py`, just below in this diff i.e. where we already do

```py
self._base_seed = _utils.worker._generate_seed(generator=loader.generator)
```

and pass all those seeds to the `_worker_loop()`. This way, all `Generators` would be explictly consumed in the `main` process (instead of just the default one), and we could just do a unified call to `generator.manual_seed(_base_seed[generator])` here.

Thoughts?

Reply:
I don't really have a strong preference on way or another. My primary constraint is that the reseeded generators should be deterministically reseeded based on the generator seed and ONLY the generator seed. Besides that, as long as your choice of new seed isn't too likely to conflict, any way of doing it is fine to me.

Reply:
> the reseeded generators should be deterministically reseeded based on the **generator** seed and ONLY the **generator** seed

If by the **generator** in bold you're referring to the one passed as parameter to `Dataloader(..., generator=generator)`, then I think that's possible - I'll try tomorrow.

If you want each generator `g` to be reseeded with a seed that only depend on `g`, then unfortunately that's not even the case right now in `main`, because of the existence of the `generator` parameter to the `Dataloader` :/

Reply:
Ah, from this and your other comment, I now understand your problem. I think this is a matter of defining semantics: do we expect the generator parameter passed to Dataloader to affect all the other generators or not? I feel there's a reasonable case to be made that it should only affect the main generator (thus the special case; I think this is what your code does now.)